hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/safemode/HealthyPipelineSafeModeRule.java

```diff
@@ -47,43 +39,51 @@
  * through in a cluster.
  */
 public class HealthyPipelineSafeModeRule
-    extends SafeModeExitRule<PipelineReportFromDatanode>{
+    extends SafeModeExitRule<Pipeline>{
 
   public static final Logger LOG =
       LoggerFactory.getLogger(HealthyPipelineSafeModeRule.class);
-  private final PipelineManager pipelineManager;
   private int healthyPipelineThresholdCount;
   private int currentHealthyPipelineCount = 0;
-  private final Map<PipelineID, Boolean> processedPipelines = new HashMap<>();
   private final double healthyPipelinesPercent;
 
   HealthyPipelineSafeModeRule(String ruleName, EventQueue eventQueue,
       PipelineManager pipelineManager,
       SCMSafeModeManager manager, Configuration configuration) {
     super(manager, ruleName, eventQueue);
-    this.pipelineManager = pipelineManager;
     healthyPipelinesPercent =
         configuration.getDouble(HddsConfigKeys.
                 HDDS_SCM_SAFEMODE_HEALTHY_PIPELINE_THRESHOLD_PCT,
             HddsConfigKeys.
                 HDDS_SCM_SAFEMODE_HEALTHY_PIPELINE_THRESHOLD_PCT_DEFAULT);
 
+    int minDatanodes = configuration.getInt(
+        HddsConfigKeys.HDDS_SCM_SAFEMODE_MIN_DATANODE,
+        HddsConfigKeys.HDDS_SCM_SAFEMODE_MIN_DATANODE_DEFAULT);
+
+    // We only care about THREE replica pipeline
+    int minHealthyPipelines = minDatanodes /
+        HddsProtos.ReplicationFactor.THREE_VALUE;
+
     Preconditions.checkArgument(
         (healthyPipelinesPercent >= 0.0 && healthyPipelinesPercent <= 1.0),
         HddsConfigKeys.
             HDDS_SCM_SAFEMODE_HEALTHY_PIPELINE_THRESHOLD_PCT
             + " value should be >= 0.0 and <= 1.0");
 
-    // As we want to wait for 3 node pipelines
-    int pipelineCount =
+    // We want to wait for RATIS THREE factor write pipelines
+    int pipelineCount = pipelineManager.getPipelines(
+        HddsProtos.ReplicationType.RATIS, HddsProtos.ReplicationFactor.THREE,
+        Pipeline.PipelineState.OPEN).size() +
         pipelineManager.getPipelines(HddsProtos.ReplicationType.RATIS,
-            HddsProtos.ReplicationFactor.THREE).size();
+            HddsProtos.ReplicationFactor.THREE,
+            Pipeline.PipelineState.ALLOCATED).size();
```

Comment:
should we consider only OPEN_RATIS_THREE pipeline here as ALLOCATED_RATIS_THREE may not finish pipeline creation.

Reply:
I think here when SCM starts on fresh cluster this count will be zero, and on already setup cluster, when SCM restarts, all pipelines will be in ALLOCATED state, once pipeline report is received, pipeline report handler moves it to OPEN state. 

I think we should consider AllocatedPipelines during pipelineCount.

Let me know if something I am missing here.

Reply:
I found a problem when the pipeline is not moved from ALLOCATED to OPEN before the test execution: https://issues.apache.org/jira/browse/HDDS-2606 

Not so familiar with this part of the code (pipeline creation) but would be great to have a check if at least one RATIS.THREE pipeline is open.  In this case it would be enough to wait until the SCM is moved out from the safe mode.

Reply:
Previously there is a problem on fresh installed cluster when pipelineAvailabilityCheck is set to true, and no pipeline will be in the system, and we come out of safeMode. (And also without HDDS-2034 the default is set to false). With this patch that issue is addressed. 

So, to avoid the issue, we need to set hdds.scm.safemode.min.datanode=3 on a 3 node cluster, so that when SCM comes out of the safe mode, writes will succeed. And also on docker-compose clusters, we need to wait until we come out of SCM safe mode, not until required number of data nodes are registered.

Reply:
Agree: "we need to wait until we come out of SCM safe mode, not until required number of data nodes are registered." This will close to the requirement of a production environment.

Reply:
Make sense. Will fix it

Reply:
> So, to avoid the issue, we need to set hdds.scm.safemode.min.datanode=3 on a 3 node cluster, so that when SCM comes out of the safe mode, writes will succeed.

I am not sure if `hdds.scm.safemode.min.datanode=3` is enough as the pipeline creation is an async operation handled by the `BackgroundPipelineCreator`. It's possible that you have 3 datanodes but the Ratis.THREE pipeline has not yet been created. At least this is what I have seen in https://issues.apache.org/jira/browse/HDDS-2606

> And also on docker-compose clusters, we need to wait until we come out of SCM safe mode, not until required number of data nodes are registered.

I agree with this part. This patch contains a shell based check to be sure that scm is not in safe mode and HDDS-2608 will provide a java based command for the same.